File: web/src/gf_apps/gf_publisher/ts/gf_post/gf_post.ts
/*
GloFlow application and media management/publishing platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

///<reference path="../../../../d/jquery.d.ts" />

import * as gf_sys_panel       from "./../../../../gf_core/ts/gf_sys_panel";
import * as gf_post_image_view from "./gf_post_image_view";
import * as gf_tagger_input_ui from "./../../../gf_tagger/ts/gf_tagger_client/gf_tagger_input_ui";

//-----------------------------------------------------
$(document).ready(()=>{
    //-------------------------------------------------
    function log_fun(p_g,p_m) {
        var msg_str = p_g+':'+p_m
        //chrome.extension.getBackgroundPage().console.log(msg_str);

        switch (p_g) {
            case "INFO":
                console.log("%cINFO"+":"+"%c"+p_m,"color:green; background-color:#ACCFAC;","background-color:#ACCFAC;");
                break;
            case "FUN_ENTER":
                console.log("%cFUN_ENTER"+":"+"%c"+p_m,"color:yellow; background-color:lightgray","background-color:lightgray");
                break;
        }
    }
    //-------------------------------------------------
    init(log_fun);
});
//-----------------------------------------------------
export function init(p_log_fun) {
    p_log_fun('FUN_ENTER', 'gf_post.init()');
    
    gf_sys_panel.init(p_log_fun);

    const post_title_str :string      = $('#post_title').text();
    const post_tags_container_element = $('#post_tags_container');
    
    //------------------------------
    //INIT IMAGE TAGGING
    $('.post_element_image').each((p_i, p_post_element)=>{
        
        const image_element          = $(p_post_element).find('img');
        const img_url_str  :string   = $(image_element).attr('src');
        const path_lst     :string[] = img_url_str.split('/'); //Uri.parse(img_url_str).pathSegments;
        const img_file_str :string   = path_lst[path_lst.length-1];
        const tags_num_int :number   = $(p_post_element).find('.tags_container .gf_post_element_tag').length;
        
        //img_file_str example - "6c4a667457f05939af6a5f68690d0f55_thumb_medium.jpeg"
        const img_id_str :string = img_file_str.split('_')[0];
        p_log_fun('INFO', 'img_id_str - '+img_id_str);

        var tag_ui_added_bool :boolean = false;
        gf_tagger_input_ui.init_tag_input(img_id_str, //p_obj_id_str
            'image',    //p_obj_type_str
            p_post_element,
            //p_onTagsCreated_fun
            (p_added_tags_lst :string[])=>{
                view_added_tags(p_post_element, p_added_tags_lst, p_log_fun);
            },
            //p_onTagUIAdd_fun
            ()=>{
                tag_ui_added_bool = true;
            },
            //p_onTagUIRemove_fun
            ()=>{
                tag_ui_added_bool = false;
            },
            p_log_fun);

        gf_post_image_view.init(p_post_element, p_log_fun);

        $(p_post_element).on('mouseenter', (p_event)=>{

            //IMPORTANT!! - only show tags_container if there are tags attached to this post_element
            if (tags_num_int > 0) {
                $(p_post_element).find('.tags_container').css('visibility', 'visible');
            }
        });
        $(p_post_element).on('mouseleave', (p_event)=>{

            //hide the tags_container only if the tagging UI is not open. 
            //if it is open we want the tags_container visible so that we can 
            //see the tags as they're added
            if (!tag_ui_added_bool) {
                $(p_post_element).find('.tags_container').css("visibility", 'hidden');
            }
        });
    });
    //------------------------------
    //VIDEO TAGGING

    $('.post_element_video').each((p_i, p_post_element)=>{

        //ADD!! - extract video ID properly
        gf_tagger_input_ui.init_tag_input('fix',
            'video',
            p_post_element,
                    
            //p_onTagsCreated_fun
            (p_added_tags_lst :string[])=>{
                view_added_tags(p_post_element, p_added_tags_lst, p_log_fun);
            },
            ()=>{}, //p_onTagUIAdd_fun
            ()=>{}, //p_onTagUIRemove_fun
            p_log_fun);
    });
    //------------------------------

    //final List<String> tags_lst = queryAll('.post_tag').map((p_element) => p_element.text);
    //gf_post_tag_mini_view.init_tags_mini_view(tags_lst,
    //                                          p_log_fun);
}
//-----------------------------------------------------
function view_added_tags(p_post_element,
    p_added_tags_lst :string[],
    p_log_fun) {
    p_log_fun('FUN_ENTER', 'gf_post.view_added_tags()');

    const tags_container_element = $(p_post_element).find('.tags_container');

    for (var tag_str of p_added_tags_lst) {
        const tag_url_str :string = '/tags/view_objects?tag='+tag_str+'&otype=image';
        const new_tag_ui_element  = $('<a class="gf_post_element_tag">'+tag_str+'</a>');
        $(new_tag_ui_element).attr('href',tag_url_str);

        //IMPORTANT!! - add the new tag link to the DOM
        $(tags_container_element).append(new_tag_ui_element);
    }
}
//-----------------------------------------------------
function get_post_element_tags_num(p_log_fun) {
    p_log_fun('FUN_ENTER', 'gf_post.get_post_element_tags_num()');
    
}

File: go/gf_apps/gf_publisher_lib/gf_publisher_core/gf_post.go
/*
GloFlow application and media management/publishing platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

package gf_publisher_core

import (
	"fmt"
	"time"
	// "github.com/globalsign/mgo/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"github.com/gloflow/gloflow/go/gf_core"
	"github.com/gloflow/gloflow/go/gf_apps/gf_images_lib/gf_images_core"
)

//------------------------------------------------

type GFpost struct {
	Id                  primitive.ObjectID `bson:"_id,omitempty"`
	IDstr               string         `bson:"id_str"`
	Tstr                string         `bson:"t"`                // "post"
	DeletedBool         bool           `bson:"deleted_bool"`
	ClientTypeStr       string         `bson:"client_type_str"`  // "gchrome_ext" //type of the client that created the post
	TitleStr            string         `bson:"title_str"`
	DescriptionStr      string         `bson:"description_str"`
	CreationDatetimeStr string         `bson:"creation_datetime_str"`
	PosterUserNameStr   string         `bson:"poster_user_name_str"` // user-name of the user that posted this post

	//------------
	// GF_IMAGES
	ThumbnailURLstr string                     `bson:"thumbnail_url_str"` // SYMPHONY 0.3
	ImagesIDsLst    []gf_images_core.GFimageID `bson:"images_ids_lst"`
	
	//------------
	PostElementsLst []*GFpostElement `bson:"post_elements_lst"`
	
	//------------
	// TAGGING

	TagsLst  []string      `bson:"tags_lst"`
	NotesLst []*GFpostNote `bson:"notes_lst"` // SYMPHONY 0.3 - notes are chunks of text (for now) that can be attached to a post
	
	//------------
	// COLORS

	// every event can have multiple colors assigned to it
	ColorsLst []string `bson:"colors_lst"`

	//------------
}

type GFpostNote struct {
	UserIDstr           string `bson:"user_id_str"`
	BodyStr             string `bson:"body_str"`
	CreationDatetimeStr string `bson:"creation_datetime_str"`
}

//------------------------------------------------

func CreateNewPost(pPostInfoMap map[string]interface{}, pRuntimeSys *gf_core.RuntimeSys) (*GFpost, *gf_core.GFerror) {
	pRuntimeSys.LogFun("FUN_ENTER", "gf_post.Create_new_post()")
	pRuntimeSys.LogFun("INFO",      "pPostInfoMap - "+fmt.Sprint(pPostInfoMap))

	// IMPORTANT!! - not all posts have "tags_lst" element, check if this is fine or if should be enforced
	// assert(pPostInfoMap.containsKey("tags_lst"));
	// assert(pPostInfoMap["tags_lst"] is List);

	post_title_str := pPostInfoMap["title_str"].(string)
	
	//--------------------
	// POST ELEMENTS

	post_elements_infos_lst := pPostInfoMap["post_elements_lst"].([]interface{})
	post_elements_lst       := create_post_elements(post_elements_infos_lst, post_title_str, pRuntimeSys)
	pRuntimeSys.LogFun("INFO","post_elements_lst - "+fmt.Sprint(post_elements_lst))

	//--------------------
	// CREATION DATETIME

	// strconv.FormatFloat(float64(time.Now().UnixNano())/1000000000.0,'f',10,64)
	creation_datetime_str := time.Now().String()

	//--------------------
	// FIX!! - id_str should be a hash, just as events/images have their ID"s generated as hashes
	
	id_str := fmt.Sprintf("%s:%s", post_title_str, creation_datetime_str)

	//--------------------
	// OPTIONAL VALUES

	// THUMBNAIL_URL
	var thumbnail_url_str string
	if _, ok := pPostInfoMap["thumbnail_url_str"]; !ok {
		thumbnail_url_str = ""
	} else {
		thumbnail_url_str = pPostInfoMap["thumbnail_url_str"].(string)
	}

	// IMAGES_IDS
	var images_ids_lst []gf_images_core.Gf_image_id
	if _, ok := pPostInfoMap["images_ids_lst"]; !ok {

		// "images_ids_lst" key was not present
		images_ids_lst = []gf_images_core.Gf_image_id{}
	} else if ids_lst,ok := pPostInfoMap["images_ids_lst"].([]gf_images_core.Gf_image_id); !ok {
		// if pPostInfoMap is coming from mongodb, its of []interface{} type, so a "type conversion" is done)
		images_ids_lst = []gf_images_core.Gf_image_id(ids_lst)
	} else {
		images_ids_lst = ids_lst
	}

	//-------------------------
	// TAGS
	var tagsLst []string

	if _, ok := pPostInfoMap["tags_lst"]; !ok {
		// if "tags_lst" key is missing, assign an empty string
		tagsLst = []string{}
	} else if input_tags_lst,ok := pPostInfoMap["tags_lst"].([]string); ok {
		// "tags_lst" is present in pPostInfoMap and is of type []string
		tagsLst = input_tags_lst

		// //CAITION!! - if pPostInfoMap is coming from mongodb, its of []interface{} type, so a "type conversion" is done.
		// //            is this ever coming from mongodb? posts are deserialized by the mgo lib, not in this function ever. 
		// tags_lst = []string(tags_lst)
	} else {
		// "tags_lst" is not of type []string
		tagsLst = pPostInfoMap["tags_lst"].([]string)
	}

	//-------------------------
	// NOTES
	var notesLst []*GFpostNote
	if _, ok := pPostInfoMap["notes_lst"]; !ok {
		notesLst = []*GFpostNote{}
	} else {
		notesInfosLst := pPostInfoMap["notes_lst"].([]map[string]interface{})
		notesLst       = createPostNotes(notesInfosLst, pRuntimeSys)
	}

	//-------------------------
	// COLORS
	var colors_lst []string
	if _,ok := pPostInfoMap["colors_lst"]; !ok {
		colors_lst = []string{}
	} else if tagsLst, ok := pPostInfoMap["colors_lst"].([]string); !ok {
		//if pPostInfoMap is coming from mongodb, its of []interface{} type, so a "type conversion" is done)
		colors_lst = []string(tagsLst)
	} else {
		colors_lst = pPostInfoMap["colors_lst"].([]string)
	}

	//--------------------
	post := &GFpost{
		IDstr:               id_str,
		Tstr:                "post",
		DeletedBool:         false,
		ClientTypeStr:       pPostInfoMap["client_type_str"].(string),
		TitleStr:            post_title_str,
		DescriptionStr:      pPostInfoMap["description_str"].(string),
		CreationDatetimeStr: creation_datetime_str,
		PosterUserNameStr:   pPostInfoMap["poster_user_name_str"].(string),
		ThumbnailURLstr:     thumbnail_url_str,
		ImagesIDsLst:        images_ids_lst,
		PostElementsLst:     post_elements_lst,
		TagsLst:             tagsLst,
		NotesLst:            notesLst,
		ColorsLst:           colors_lst,
	}
	return post, nil
}

//------------------------------------------------	
// a post has to first be created, and only then can it be published

func publish(p_post_title_str string, pRuntimeSys *gf_core.RuntimeSys) {
	pRuntimeSys.LogFun("FUN_ENTER", "gf_post.publish()")




	


}

//------------------------------------------------
func createPostNotes(p_raw_notes_lst []map[string]interface{}, pRuntimeSys *gf_core.RuntimeSys) []*GFpostNote {

	notes_lst := []*GFpostNote{}
	for _, noteMap := range p_raw_notes_lst {
		
		snippet := &GFpostNote{
			UserIDstr: "anonymous",
			BodyStr:   noteMap["body_str"].(string),
		}
		notes_lst = append(notes_lst, snippet)
	}
	return notes_lst
}

//------------------------------------------------
func GetPostURL(pPostTitleStr string) string {



	return ""


}